src/server/namespace.cc

```diff
@@ -0,0 +1,175 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+
+#include "namespace.h"
+
+#include "jsoncons/json.hpp"
+
+// Error messages
+constexpr const char* kErrNamespaceExists = "the namespace already exists";
+constexpr const char* kErrTokenExists = "the token already exists";
+constexpr const char* kErrNamespaceNotFound = "the namespace was not found";
+constexpr const char* kErrRequiredPassEmpty = "forbidden to add namespace when requirepass was empty";
+constexpr const char* kErrClusterModeEnabled = "forbidden to add namespace when cluster mode was enabled";
+constexpr const char* kErrDeleteDefaultNamespace = "forbidden to delete the default namespace";
+constexpr const char* kErrAddDefaultNamespace = "forbidden to add the default namespace";
+constexpr const char* kErrInvalidToken = "the token is duplicated with requirepass or masterauth";
+
+Status IsNamespaceLegal(const std::string& ns) {
+  if (ns.size() > UINT8_MAX) {
+    return {Status::NotOK, fmt::format("size exceed limit {}", UINT8_MAX)};
+  }
+  char last_char = ns.back();
+  if (last_char == std::numeric_limits<char>::max()) {
+    return {Status::NotOK, "namespace contain illegal letter"};
+  }
+  return Status::OK();
+}
+
+Status Namespace::Load() {
+  auto config = storage_->GetConfig();
+  // Load from the configuration file first
+  tokens_ = config->load_tokens;
+
+  // We would like to load namespaces from db even if repl_namespace_enabled is false,
+  // this can avoid missing some namespaces when turn on/off repl_namespace_enabled.
+  std::string value;
+  auto s = storage_->Get(rocksdb::ReadOptions(), cf_, kNamespaceDBKey, &value);
+  if (!s.ok()) {
+    if (s.IsNotFound()) return Status::OK();
+    return {Status::NotOK, s.ToString()};
+  }
+  jsoncons::json j = jsoncons::json::parse(value);
+  for (const auto& iter : j.object_range()) {
+    if (tokens_.find(iter.key()) == tokens_.end()) {
+      // merge the namespace from db
+      tokens_[iter.key()] = iter.value().as<std::string>();
+    }
+  }
+  return rewriteOrWriteDB();
+}
+
+StatusOr<std::string> Namespace::Get(const std::string& ns) const {
```

Comment:
I think if the call frequency of `Get`, `Set` ... is not so high, we can just get or set it directly in the storage via rocksdb api to avoid `Load` and `rewriteOrWriteDB` in many places? (after using a schema like https://github.com/apache/kvrocks/pull/1776/files#r1342042828)

Reply:
We need to check if the namespace/token is correct when authenticating, so we need to keep them in memory.

Reply:
Got it. So you mean that it will affect the performance of the authentication if it is not in memory?

Reply:
Yes, I guess it will affect the performance a lot if the users connect-write-close every time without keeping the connection.